scalyr_agent/builtin_monitors/syslog_monitor.py

```diff
@@ -584,27 +615,205 @@ def process(self, data, handle_frame):
                 limit_key="syslog-no-frames",
             )
 
-        self._remaining = self._remaining[self._offset :]
+
+class SyslogRawRequestParser(SyslogRequestParser):
+    """
+    Special request parser which doesn't perform any handling of the received data, but writes it
+    as-is to a file on disk (aka sends it to Scalyr logger class).
+
+    It means it doesn't handle framed messages and received lines won't always be written as a
+    complete atomic unit to a file on disk at once, but as part of multiple write calls.
+
+    This handler is to be used when we want to avoid expensive framed message parsing and just want
+    to write received data as-is. It's much more efficient and offers much better throughput than
+    the default parser which handles framed data, etc.
+    """
+
+    def process(self, data, handle_frame):
+        handle_frame(data)
+
+
+class SyslogBatchedRequestParser(SyslogRequestParser):
+    """
+    This parser works in exactly the same manner as the default request parser (it supports framed
+    and line delimited data), but instead of calling scalyr logger class and writing each frame /
+    line as it's processed, it writes all the processed lines in a single batch at the end of
+    processing of a specific TCP packet.
```

Comment:
So, if I'm understanding this correctly, we are also not (really) increasing how much data we may drop if the Scalyr Agent shuts down unexpectantly, correct?  I could argue that if we write each line to disk one at time, then those lines become durable more quickly.  But, here, you are still guaranteeing all bytes are on disk when processing a single glob of data from TCP/IP.  We aren't waiting for another packet to be received, some time out to be exceeded (beyond the incomplete frame thing), etc that would significantly increase exposure.

Is that correct?

Heck, in syslog, we aren't really even acknowledging that we have received the bytes, are we?  We don't send a response.  So, syslog data could be lost in any point from buffering on the client's TCP/IP side to data in transit on the network to network buffering on the Scalyr Agent side.  Is that right?

Reply:
Correct, yeah.

Current code is not ideal in terms of the how much control we have of exactly when (e.g. after how many bytes / lines received) we actually flush data to disk / logger, but changing that would require bigger refactor and logic change.

And I don't think that change would bring us any substantial improvement in terms of throughput itself (with this change it's already good enough aka syslog monitor won't be the bottleneck), we would just get more control over "durability" and potentially the data loss window in a case of hard crash would be slightly smaller.

So for now, this change is likely fine, but going forward we may want to change the logic a bit to give us more control over the "flush" interval.

As far as actual numbers of how many times ``handle_frame`` was called before and with this change - previously, if we sent 13k messages it would be called 13k times (once per message) and now it will be called approximately total_data_sent_in_bytes / tcp_buffer_size which will be substantially less than 13k times.

Reply:
Ok, thanks for the explanation and yup, this approach looks good to me.  We can touch base on Monday morning, but yeah, we may very will use an internal build with this change for a customer looking for higher throughput from syslog.